socket.1
.\" -*- nroff -*-
.ig
This file is part of Socket-1.5.

Copyright (c) 1992 - 2018 Juergen Nickelsen <ni@w21.org>
and Boris Nikolaus. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.

     $Id$
..
.TH SOCKET 1 "Apr 08. 2007"
.SH NAME
socket \- create tcp socket and connect to stdin/out
.SH SYNOPSIS
.B socket
[
.B \-bcefhHnqQrvvw46
]
[
.B \-a
.I bind-address
]
[
.B \-p
.I command
]
[
.B \-T
.I timeout
]
.I host port
.br
.B socket
[
.B \-46bcefhHlnqQrRvvw
]
[
.B \-a
.I bind-address
]
[
.B \-p
.I command
]
[
.B \-T
.I timeout
]
.B \-s 
[
.B \-l
]
.I port
.SH DESCRIPTION
.B Socket
creates an Internet domain TCP socket and connects it to stdin and stdout.
The 
.I host
argument can be an Internet number in dot-notation (like
``130.149.28.10'') or a domain name. In this case it must be possible
to resolve the name to a valid Internet address with
.IR gethostbyname (3).
The 
.I port
argument can be a port number or a service name which can be mapped to
a port number by
.IR getservbyname (3).
.SH OPTIONS
.TP
.BI "\-a " bind-address
Bind the socket to the specified
.I bind-address
instead of the default address (in client mode) or all addresses of
the host (in server mode).
This may lead to a connection failure in client mode if there is no
route from the
.I bind-address
to the destination address.
.TP
.BR "\-4 " (IPv4)
Only use IP protocol version 4. (Default: Support both IPv4 and IPv6
if IPv6 support has been compiled in).
.TP
.BR "\-6 " (IPv6)
Only use IP protocol version 6 (if support has been compiled in).
(Default: Support both IPv4 and IPv6 if IPv6 support has been compiled
in).
.TP
.BR "\-b " (background)
The program forks itself into the background, detaches from its
controlling tty, closes the file descriptors associated with the tty,
and changes its current directory to the root directory.
.TP
.BR "\-c " (crlf)
Linefeed characters (LF) are converted to a Carriage Return Linefeed
sequence (CRLF) when written to the socket.
CRLF sequences read from the socket are converted to a single LF.
.TP
.BR "\-e " "(no stderr capture)"
With -p: stderr from the called program is not redirected to the
socket, but sent to the Socket program's original stderr.
.TP
.BR "\-f " (fork)
When a server connection has been accepted, a separate process is
forked to handle the connection in background.
.TP
.BR "\-h " (help)
Display the usage of the program.
.TP
.BR "\-H " (half-close)
Enable half-close support. This allows to close both directions of the
connection separately by sending an end-of-file indicator for each
direction. The connection is assumed alive until both directions have
been closed.
.TP
.BR "\-l " (loop)
(only valid with
.BR \-s )
After a connection has been closed,
another connection is accepted.
.TP 
.BR "\-n " "(no reverse lookup)"
Don't do reverse lookup of peer addresses.
.TP
.BR "\-p " (program)
The specified 
.I command
is executed for each connection. Its standard input, standard output,
and standard error channels are connected to the socket.
The environment variables SOCKET_LOCAL_ADDRESS and SOCKET_LOCAL_PORT
are set to the socket's local IP address and port number; the
environment variables SOCKET_PEER_ADDRESS and SOCKET_PEER_PORT are
correspondingly set to IP address and port number of the socket on
the peer's side.
.I Command
can be any shell command since it is passed to \fC/bin/sh\fP for
execution.
.TP
.BR "\-q " (quit)
The connection is closed when an end-of-file condition occurs on standard
input.
.TP
.BR "\-Q " "(reset on ungraceful quit)"
When the program is terminated by SIGINT or SIGTERM or if reading from
stdin or writing to stdout fails, the connection is closed
ungracefully by sending a TCP RESET instead of TCP FIN. This allows to
detect the abortion at the remote side where
.IR read (2)
returns a "Connection reset by peer" error.
.TP
.BR "\-r " "(read only)"
No data is read from standard input and written to the socket.
.TP
.BR "\-R " "(no REUSEADDR)"
Don't set the socket option SO_REUSEADDR.
.TP
.BR "\-s " (server)
A server socket is created.
A
.I hostname 
argument is not required.
.TP
.BR "\-T " (timeout)
The specified
.I timeout
(in seconds) is applied to the
.IR connect (2)
and
.IR select (2)
system calls, such that the program is terminated when that timeout
expires on connection setup or when reading from the socket or
standard input.
.TP
.BR "\-v " (verbose)
Each ``v'' in the options increases the verbosity level by one.
Currently supported are 
.BR 1 :
print information about connections, and 
.BR 2 :
additionally print information about name resolution and connection
attempts.
.TP
.BR "\-w " "(write only)"
No data is read from the socket and written to the standard output.
.TP
.B \-version
.B Socket
prints its version ID and terminates. 
This must be the first argument to have an effect.
.SH EXAMPLES
The command
.IP
\fCsocket -v coma.cs.tu-berlin.de nntp\fP
.LP
connects to the nntp port (port 119) of coma.cs.tu-berlin.de
(130.149.28.10).
.br
The command
.IP
\fCsocket \-sl 3425\fP
.LP
creates a server socket on port 3425 on the local host and waits for a
connection.
After a connection has been closed, a new connection is accepted. 
.br
The command
.IP
\fCsocket \-wslqvp "echo Socket! " 1938\fP
.LP
creates a server socket on port 1938 on the local host and waits for a
connection. When a connection is accepted, the string "Socket!" is
written to the socket. No data is read from the socket and written to
the \fCfinger\fP program. The connection is closed when an end-of-file
condition at the standard output of the program occurs. Then a new
connection is accepted.
.SH DIAGNOSTICS
Lots of diagnostics for failed system calls.
.TP
.BI "unknown host " host
.IR host 's
address could not be resolved.
.B socket
terminates due to an error condition.
.SH SEE ALSO
.BR accept (2),
.BR bind (2),
.BR connect (2), 
.BR listen (2), 
.BR socket (2),
.BR getaddrinfo (3),
.BR gethostbyname (3),
.BR getservbyname (3)
.SH BUGS
The version 2 release is overdue.
.LP
\fCsocket \-p\fP terminates due to a SIGPIPE signal if there is more
data from the socket available than the child process wants to read.
.LP
Please report any other bugs to the author.
.SH VERSION
This manual page describes Socket\-1.5.
.SH AUTHOR
Juergen Nickelsen <ni@w21.org>
.br
Boris Nikolaus made some significant changes in version 1.5, most
notably for IPv6.


CHANGES
-*-text-*-

Changes in Socket version 1.5p3

  - updated years in COPYRIGHT
  - new example script mconnect
  - extended build info in binary
  - silenced some compiler warnings

Changes in Socket version 1.5p2

  - updated author email address
  - removed code for platforms without IPv6 support
  - some minor cleanup

Changes in Socket version 1.5p1

  - works on IRIX (6.5) again

Changes in Socket version 1.5

  - exit gracefully on SIGHUP

  by Boris Nikolaus:

  - -H option for half-close support of TCP connections

  - -Q option for TCP RESET on SIGINT and SIGTERM

  - IPv6 support added, -4 and -6 options for manual selection of
    protocol version

  - -h option for usage

  - a few bug fixes

  other changes:

  - build support for Interix (Microsoft Windows Services For Unix)

Changes in Socket version 1.4:

  - no more tab characters in the sources

  - support for building multiple architectures in one directory;
    requires GNU Make

  - -e option for not redirecting stderr of a called program to the
    socket, by Robert Lemmen

Changes in Socket version 1.3:

  - Fixed bug related to partial writes: always writing from the
    beginning of the block after partial writes caused data
    corruption after write(2) had been interrupted by a signal.

  - Fixed bug for 64-bit platforms: Change type of IP addresses from
    unsigned long to uint32_t.

  - Change of author e-mail address

  - -n option for not doing reverse lookups

  - -a bind-address option to bind the socket to a specific local IP
    address

  - socket address environment variables for child process

  - fixed bug with accept() and getpeername() addrlen argument type,
    which triggered a gcc warning

  - some minute documentation fixes

Changes in Socket version 1.2p1:

  - Fixed bug with incorrectly initialized SIGCHLD handler

  - ISOfied the remaining functions

  - Some other small style fixes

Changes in Socket version 1.2:

  - Adapted to Linux (in this case RH 5.2 with kernel 2.2.2), Solaris
    2.6 / 7, and FreeBSD 2.2.7. Other platforms are not (explicitly)
    supported by this release.

  - Except for SIGCHLD no signals are handled specially; siglist.c
    removed.

  - Multiple -vv options increase verbosity.

  - Sets SO_REUSEADDR unless -R option is specified.

  - Needs an ISO C compiler (and I hope it is not GCC dependent).

  - License changed to Berkeley-style.

  - Moved from ShapeTools to CVS (and a new repository -- that all
    source version numbers are 1.2 is *really* coincidental!)



socket.c
/* This file is part of Socket-1.5.
 */

/*-
 * Copyright (c) 1992 - 2018 Juergen Nickelsen <ni@w21.org>
 * and Boris Nikolaus. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *      $Id$
 */


#include "globals.h"

#include <netdb.h>
#include <errno.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <netinet/in.h>
#ifdef SEQUENT
#include <strings.h>
#else
#include <string.h>
#endif

/* global variables */
int forkflag = 0 ;              /* server forks on connection */
int serverflag = 0 ;            /* create server socket */
int loopflag = 0 ;              /* loop server */
int nostderrflag = 0 ;          /* don't write stderr to socket */
int verboseflag = 0 ;           /* give messages */
int readonlyflag = 0 ;          /* only read from socket */
int writeonlyflag = 0 ;         /* only write to socket */
int quitflag = 0 ;              /* quit connection on EOF */
int crlfflag = 0 ;              /* socket expects and delivers CRLF */
int backgflag = 0 ;             /* put yourself in background */
int noreverseflag ;             /* don't do reverse lookup of peer addresses */
int active_socket ;             /* socket with connection */
int Reuseflag = 1 ;             /* set server socket SO_REUSEADDR */
int halfcloseflag = 0 ;         /* support half-close of sockets */
int resetflag = 0 ;             /* reset socket on SIGINT or SIGTERM */
int protocol_family = AF_UNSPEC ;/* protocol family to use */
unsigned int timeout = 0 ;      /* timeout in seconds */
char *progname ;                /* name of the game */
char *pipe_program = NULL ;     /* program to execute in two-way pipe */
char *bind_address = NULL ;     /* address to listen on */
int running = 1;                /* non-zero while running, reset on SIGHUP */

jmp_buf setjmp_env ;            /* buffer for return via longjmp() */
int alarmsig_occured = 0 ;      /* non-zero after alarm signal */

int server(char *service) ;
int handle_server_connection(void) ;
int client(char *host, char *service) ;

int main(int argc, char *argv[])
{
    char *cp ;                  /* to point to '/' in argv[0] */
    int opt ;                   /* option character */
    int error = 0 ;             /* usage error occurred */
    extern int optind ;         /* from getopt() */
    char *service ;             /* service (name or port) for socket */
    int ipv4 = 0 ;              /* IPv4 requested */
    int ipv6 = 0 ;              /* IPv6 requested */
    int exitval ;               /* Exit code of the program */

    /* print version ID if requested */
    if (argv[1] && !strcmp(argv[1], "-version")) {
        puts(so_release()) ;
        exit(0) ;
    }

    /* set up progname for later use */
    progname = argv[0] ;
    if ((cp = strrchr(progname, '/'))) {
        progname = cp + 1 ;
    }

    /* parse options */
    while ((opt = getopt(argc, argv, "a:bcefHlnp:qQrRsvwT:46h?")) != -1) {
        switch (opt) {
          case 'n':
            noreverseflag = 1 ;
            break ;
          case 'a':
            bind_address = optarg ;
            break ;
          case 'f':
            forkflag = 1 ;
            break ;
          case 'c':
            crlfflag = 1 ;
            break ;
          case 'w':
            writeonlyflag = 1 ;
            break ;
          case 'e':
            nostderrflag = 1 ;
            break ;
          case 'p':
            pipe_program = argv[optind - 1] ;
            break ;
          case 'q':
            quitflag = 1 ;
            break ;
          case 'r':
            readonlyflag = 1 ;
            break ;
          case 'R':
            Reuseflag = 0 ;
            break ;
          case 's':
            serverflag = 1 ;
            break ;
          case 'v':
            verboseflag += 1 ;
            break ;
          case 'l':
            loopflag = 1 ;
            break ;
          case 'b':
            backgflag = 1 ;
            break ;
          case 'T':
            timeout = atoi(optarg) ;
            break ;
          case 'H':
            halfcloseflag = 1 ;
            break ;
          case 'Q':
            resetflag = 1 ;
            break ;
          case '4':
            ipv4 = 1 ;
            break ;
          case '6':
            ipv6 = 1 ;
            break ;
          default:
            error++ ;
        }
    }
    if (error ||                /* usage error? */
        argc - optind + serverflag != 2) { /* number of args ok? */
        usage() ;
        exit(15) ;
    }

    /* check some option combinations */
#define senseless(s1, s2) \
    fprintf(stderr, "It does not make sense to set %s and %s.\n", (s1), (s2))

    if (writeonlyflag && readonlyflag) {
        senseless("-r", "-w") ;
        exit(15) ;
    }
    if (loopflag && !serverflag) {
        senseless("-l", "not -s") ;
        exit(15) ;
    }
    if (backgflag && !serverflag) {
        senseless("-b", "not -s") ;
        exit(15) ;
    }
    if (forkflag && !serverflag) {
        senseless("-f", "not -s") ;
    }
    if (ipv4 && !ipv6) {
        protocol_family = AF_INET ;
    } else if (ipv6 && !ipv4) {
        protocol_family = AF_INET6 ;
    } else {
        protocol_family = AF_UNSPEC ;
    }

    /* set up signal handling */
    init_sighandlers() ;

    /* get service name */
    service = argv[optind + 1 - serverflag] ;

    /* and go */
    if (serverflag) {
        if (backgflag) {
            background() ;
        }
        exitval = server(service) ;
    } else {
        exitval = client(argv[optind], service) ;
    }
    exit(exitval) ;
}


int server(char *service)
{
    int socket_handle ;
    socklen_t soa_len ;
    char *host_name ;
    char *service_name ;
    char sa_buffer[256] ;
    struct sockaddr *sa = (struct sockaddr *) sa_buffer ;
    int retval = 0 ;

    /* allocate server socket */
    soa_len = sizeof(sa_buffer) ;
    if ((socket_handle = create_server_socket(bind_address, service, 1,
                                              sa, &soa_len)) < 0) {
        return 1 ;
    }
    if (verboseflag) {
        fprintf(stderr, "listening on") ;
        if (bind_address != NULL) {
            fprintf(stderr, " interface address %s", get_ipaddr(sa, soa_len)) ;
            if ((host_name = get_hostname(sa, soa_len)) != NULL) {
                fprintf(stderr, " (%s)", host_name) ;
            }
        }
        fprintf(stderr, " port %s", get_port(sa, soa_len)) ;
        if ((service_name = get_service(sa, soa_len, "tcp")) != NULL) {
            fprintf(stderr, " (%s)", service_name) ;
        }
        fprintf(stderr, "\n") ;
    }

    /* server loop */
    for (;;) {

        /* accept a connection */
        soa_len = sizeof(sa_buffer) ;
        if ((active_socket = accept(socket_handle, sa, &soa_len)) < 0) {
            if (errno == ECONNABORTED)
                continue ;
#ifdef EPROTO
            if (errno == EPROTO)
                continue ;
#endif
            if (errno != EINTR) {
                perror2("accept") ;
                retval = 3 ;
            }
        } else {
            /* if verbose, get name of peer and give message */
            if (verboseflag) {
                fprintf(stderr, "connection from %s", get_ipaddr(sa, soa_len)) ;
                if ((host_name = get_hostname(sa, soa_len)) != NULL) {
                    fprintf(stderr, " (%s)", host_name) ;
                }
                fprintf(stderr, " port %s", get_port(sa, soa_len)) ;
                if ((service_name = get_service(sa, soa_len, "tcp")) != NULL) {
                    fprintf(stderr, " (%s)", service_name) ;
                }
                fprintf(stderr, "\n") ;
            }
            if (forkflag) {
                switch (fork()) {
                  case 0:
                    retval = handle_server_connection() ;
                    exit(retval) ;
                  case -1:
                    perror2("fork") ;
                    break ;
                  default:
                    close(active_socket) ;
                    retval = wait_for_children() ;
                }
            } else {
                retval = handle_server_connection() ;
            }
        }
        if (!loopflag) {
            break ;
        }
    }
    return retval ;
}

int handle_server_connection(void)
{
    int retval ;

    /* open pipes to program, if requested */
    if (pipe_program != NULL) {
        open_pipes(pipe_program) ;
    }
    /* enter IO loop after establishing return point */
    if (!setjmp(setjmp_env)) {
        alarmsig_occured = 0 ;
        retval = do_io() ;
    } else {
        retval = 3 ;
    }
    /* connection is closed now */
    close(active_socket) ;
    if (pipe_program != NULL) {
        /* signal EOF to program */
        close(0) ;
        close(1) ;
        /* remove zombies */
        wait_for_children() ;
    }
    return retval ;
}

int client(char *host, char* service)
{
    char *host_name ;
    char *service_name ;
    char sa_buffer[256] ;
    struct sockaddr *sa = (struct sockaddr *) sa_buffer ;
    socklen_t soa_len ;
    int retval ;

    /* get connection */
    soa_len = sizeof(sa_buffer) ;
    if ((active_socket = create_client_socket(bind_address, host, service,
                                         sa, &soa_len)) < 0) {
        return 1 ;
    }
    if (verboseflag) {
        fprintf(stderr, "connected to %s", get_ipaddr(sa, soa_len)) ;
        if ((host_name = get_hostname(sa, soa_len)) != NULL) {
            fprintf(stderr, " (%s)", host_name) ;
        }
        fprintf(stderr, " port %s", get_port(sa, soa_len)) ;
        if ((service_name = get_service(sa, soa_len, "tcp")) != NULL) {
            fprintf(stderr, " (%s)", service_name) ;
        }
        fprintf(stderr, "\n") ;
    }

    /* open pipes to program if requested */
    if (pipe_program != NULL) {
        open_pipes(pipe_program) ;
    }
    /* enter IO loop */
    retval = do_io() ;
    /* connection is closed now */
    close(active_socket) ;
    return retval ;
}

/* EOF */
